lib/ocpp/v201/profile.cpp

```diff
@@ -0,0 +1,684 @@
+            const auto start_schedule = floor_seconds(in_profile.chargingSchedule.front().startSchedule.value());
+            const auto end = floor_seconds(in_end);
+            const auto now_tp = start.to_time_point();
+            int seconds_to_go_back{0};
+            int seconds_to_go_forward{0};
+
+            /*
+             example problem case:
+             - allow daily charging 08:00 to 18:00
+               at 07:00 and 19:00 what should the start time be?
+             a) profile could have 1 period (32A) at 0s with a duration of 36000s (10 hours)
+                relying on a lower stack level to deny charging
+             b) profile could have 2 periods (32A) at 0s and (0A) at 36000s (10 hours)
+                i.e. the profile covers the full 24 hours
+             at 07:00 is the start time in 1 hour, or 23 hours ago?
+             23 hours ago is the chosen result - however the profile code needs to consider that
+             a new daily profile is about to start hence the next start time is provided.
+             Weekly has a similar problem
+            */
+
+            switch (in_profile.recurrencyKind.value()) {
+            case RecurrencyKindEnum::Daily:
+                seconds_to_go_back = duration_cast<seconds>(now_tp - start_schedule.to_time_point()).count() %
+                                     (HOURS_PER_DAY * SECONDS_PER_HOUR);
+                if (seconds_to_go_back < 0) {
+                    seconds_to_go_back += HOURS_PER_DAY * SECONDS_PER_HOUR;
+                }
+                seconds_to_go_forward = HOURS_PER_DAY * SECONDS_PER_HOUR;
+                break;
+            case RecurrencyKindEnum::Weekly:
+                seconds_to_go_back = duration_cast<seconds>(now_tp - start_schedule.to_time_point()).count() %
+                                     (SECONDS_PER_DAY * DAYS_PER_WEEK);
+                if (seconds_to_go_back < 0) {
+                    seconds_to_go_back += SECONDS_PER_DAY * DAYS_PER_WEEK;
+                }
+                seconds_to_go_forward = SECONDS_PER_DAY * DAYS_PER_WEEK;
+                break;
+            default:
+                EVLOG_error << "Invalid RecurrencyKindType: " << static_cast<int>(in_profile.recurrencyKind.value());
+                break;
+            }
+
+            start = std::move(DateTime(now_tp - seconds(seconds_to_go_back)));
+
+            while (start <= end) {
+                start_times.push_back(start);
+                start = DateTime(start.to_time_point() + seconds(seconds_to_go_forward));
+            }
+        }
+        break;
+    case ChargingProfileKindEnum::Relative:
+        // if there isn't a session start then assume the session starts now
+        if (in_session_start) {
+            start = floor_seconds(in_session_start.value());
+        }
+        start_times.push_back(start);
+        break;
+    }
+    return start_times;
+}
+
+/// \brief calculate the start times for the schedule period
```

Comment:
I think this description might need an update. It looks like this method returns profile entries, not just start times. Also, am I right that it only returns multiple entries for recurrent profiles? It might be worth saying as much in the description.

Reply:
TBH, this was supposed to be removed and only be in the header file, as recommended by C++ big brain types. The docs match the 1.6 code, but agree so updated in the header file.